module/zfs/vdev.c

```diff
@@ -1317,6 +1365,11 @@ vdev_metaslab_group_create(vdev_t *vd)
 		vd->vdev_mg = metaslab_group_create(mc, vd,
 		    spa->spa_alloc_count);
 
+		if (!vd->vdev_islog) {
+			vd->vdev_log_mg = metaslab_group_create(
+			    spa_embedded_log_class(spa), vd, 1);
+		}
```

Comment:
I may have misread the flow here (e.g. we call `vdev_metaslab_group_create` before `vdev_metaslab_init`), but wouldn't the above set `vdev_log_mg` for normal vdevs, regardless of our `zfs_embedded_slog_min_ms` tunable?

This seems to be the case at least for `zdb.c`.

Reply:
That's right.  We always have the vdev_log_mg (for non-log vdevs) but it won't have any metaslabs if the vdev is too small (based on `zfs_embedded_slog_min_ms`).  That happens in `vdev_metaslab_init()`:
```c
	/*
	 * Find the emptiest metaslab on the vdev and mark it for use for
	 * embedded slog by moving it from the regular to the log metaslab
	 * group.
	 */
	if (vd->vdev_mg->mg_class == spa_normal_class(spa) &&
	    vd->vdev_ms_count > zfs_embedded_slog_min_ms &&
	    avl_is_empty(&vd->vdev_log_mg->mg_metaslab_tree)) {

```

Reply:
Ugh apologies, I think I confused myself for assuming for some reason that small vdevs are not supposed to have a log metaslab group.

That makes sense and it's great that the rest of the code doesn't need to be guarded by if statements.

🚢 🇮🇹